### DISCUSSION TITLE
Why was JJ implemented in Rust over Golang?

### DISCUSSION THREAD

**Discussion by jwendel** [jj-vcs/jj#1662] [General]
Just wondering, but why did you decide to implement this in Rust vs Golang (or C++)? Being that this is intended to be used at Google where Go is likely more widely used than Rust, I'm just wondering what the reasoning was?

**Comment by martinvonz** [comment]
It started as my personal project (that's why it's still under my GitHub account rather than under some team name). I had heard lots of good things about Rust, so I wanted to get an opportunity to try it on a real project. While I considered how we could integrate jj into Google's ecosystem from the beginning (or very close anyway - I can't remember exactly), I somehow didn't consider the impact of the language choice until much later. I'm not sure how much more I can say, but since you seem to work at Google, I'm happy to provide more detail on internal chat.

**Comment by arxanas** [comment]
[Copying from Discord](https://discord.com/channels/968932220549103686/969829516539228222/1138954559557996614):

>>    Firstly, out of curiosity: was Rust the first choice or were there some rewrites along the road?
>
> Yes, it was my first choice in that sense (that I didn't do a rewrite), but not in the sense that it was my favorite language - I had just heard good things about it and wanted to try it.
>
>>    And secondly - of personal importance to me: are there reasons for why it was Rust and not say - C/C++/Go (especially since the project started at Google as far as I understood).
>
> I hate C and C++. I've never tried Go. I wrote mostly Java before starting on Google's source control team. I got used to using IDEs and the help the provide. I know many people like C++. I think part of the reason I don't like it is that I'm forgetful and can't memorize large APIs like other people can, so I rely a lot on my IDE. When writing C++, I spend a lot of time looking for API documentation and examples in existing code.
>
> [...]
>
>> I recall that JGit has trouble dealing with large objects/buffers in a performant way because of the boxing, etc.; using Rust helps avoid that one case 🙂 (as opposed to Java and potentially Go?)
>
> Yes, Java was never an option in my mind
>
> [...]
>
>>    Though for Rust - do I understand correctly that it was just chosen out of curiosity and the popularity at that time and then you stuck with it, because it happened to cover all that was needed?
>> Or was there some more consideration for language features, etc?
>
> The former, i.e. it covered all that I needed (fast native code, good IDE support, high level of abstraction, safe).
